File: jobs/research.py
"""Research Brief from her data + web + her uploaded reports (BUILD_PLAN.md section 6).

Spec (worker/jobs/research.ts buildSpec):
  profile        her locked Brand Profile sections
  her_data       {summary: {platform: {videos, avg_views, history_days, top_times}}, accounts,
                  learned_slots, timezone}
  uploads        [{id, source_id, title, r2_key, ext}]   outside reports she uploaded
  baseline       section 10b: {slots, caps, sources}      the starting schedule and its studies
  features       {deeper_research}                        OpenRouter Perplexity search (paid, off by default)
  keys           {openrouter, firecrawl}                  from her Connect screen (env fallback)
  model, deeper_model, system, web_skipped_source_id

Result: {"body": BriefBody, "sources": BriefSource[]} exactly as shared/types.ts. Posting times
are built here from the 10b baseline (or her learned slots) so they always cite the studies;
the model writes the rest and may cite only the source ids it is given. The Worker re-applies
the truth rules (worker/domain/brief.ts) before storing the draft.
Logs: step names and counts only (section 13).
"""
from __future__ import annotations

import json
import os
import time
import urllib.error
import urllib.request
from pathlib import Path
from typing import Any

from common import WORK, Job, download_input, log, run
from extract import Unreadable, extract_text, llm_json

PLATFORMS = ("tiktok", "instagram", "youtube")
LABEL = {"tiktok": "TikTok", "instagram": "Instagram", "youtube": "YouTube Shorts"}
DAY = ["Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday"]
EXCERPT = 2500


def hour_label(h: int) -> str:
    return f"{12 if h % 12 == 0 else h % 12} {'am' if h < 12 else 'pm'}"


def claim(text: str, ids: list[str], basis: str, solid: bool = True) -> dict[str, Any]:
    return {"text": text, "source_ids": ids, "basis": basis, "confidence": "solid" if solid else "uncertain"}


# ---------- web ----------

def firecrawl_search(key: str, query: str, limit: int = 4) -> list[dict[str, str]]:
    body = json.dumps({"query": query, "limit": limit, "scrapeOptions": {"formats": ["markdown"], "onlyMainContent": True}}).encode()
    req = urllib.request.Request("https://api.firecrawl.dev/v1/search", data=body, method="POST", headers={"Authorization": f"Bearer {key}", "Content-Type": "application/json"})
    for attempt in range(3):
        try:
            with urllib.request.urlopen(req, timeout=120) as res:
                data = json.loads(res.read().decode())
            out = []
            for item in data.get("data") or []:
                url = item.get("url") or (item.get("metadata") or {}).get("sourceURL")
                if not url:
                    continue
                out.append({"url": url, "title": (item.get("title") or (item.get("metadata") or {}).get("title") or url)[:200], "text": (item.get("markdown") or item.get("description") or "")[:EXCERPT]})
            return out
        except urllib.error.HTTPError as e:
            log("firecrawl.retry", status=e.code, attempt=attempt)
            if e.code in (401, 402, 403):
                return []
        except (urllib.error.URLError, TimeoutError, json.JSONDecodeError):
            log("firecrawl.retry", attempt=attempt)
        time.sleep(3 * 2 ** attempt)
    return []


def queries_from_profile(profile: dict[str, str] | None) -> list[str]:
    qs = [
        "short-form video hooks that work for women over 50 lifestyle creators",
        "luxury event host content ideas for Instagram Reels and TikTok",
        "best video length for Instagram Reels and TikTok engagement 2026 study",
    ]
    themes = (profile or {}).get("themes", "")
    for line in themes.splitlines():
        t = line.strip("•-* ").split(":")[0].strip()
        if 3 < len(t) < 80:
            qs.append(f"{t} short video ideas")
        if len(qs) >= 6:
            break
    return qs


def deeper_search(key: str, model: str, profile: dict[str, str] | None) -> tuple[str, list[dict[str, str]]]:
    """OpenRouter Perplexity search (about half a cent a search); returns text + cited pages."""
    q = "What short-form video formats, hooks and posting habits work best for a luxury event and lifestyle creator whose audience is mostly women over 40? Cite studies."
    if profile and profile.get("who"):
        q += " Creator: " + profile["who"][:600]
    body = json.dumps({"model": model, "messages": [{"role": "user", "content": q}], "max_tokens": 1200}).encode()
    req = urllib.request.Request("https://openrouter.ai/api/v1/chat/completions", data=body, method="POST", headers={"Authorization": f"Bearer {key}", "Content-Type": "application/json", "X-Title": "Sheila Studio"})
    try:
        with urllib.request.urlopen(req, timeout=180) as res:
            data = json.loads(res.read().decode())
    except Exception:  # noqa: BLE001
        log("deeper.failed")
        return "", []
    msg = (data.get("choices") or [{}])[0].get("message") or {}
    pages: list[dict[str, str]] = []
    for a in msg.get("annotations") or []:
        u = (a.get("url_citation") or {})
        if u.get("url"):
            pages.append({"url": u["url"], "title": (u.get("title") or u["url"])[:200]})
    for u in data.get("citations") or []:
        if isinstance(u, str) and all(p["url"] != u for p in pages):
            pages.append({"url": u, "title": u[:200]})
    return (msg.get("content") or "")[:EXCERPT * 2], pages[:6]


# ---------- posting times (always cited) ----------

def best_times(spec: dict[str, Any], her_ids: set[str]) -> dict[str, list[dict[str, Any]]]:
    learned = spec.get("her_data", {}).get("learned_slots") or {}
    slots = spec["baseline"]["slots"]
    out: dict[str, list[dict[str, Any]]] = {}
    for p in PLATFORMS:
        rows = []
        if learned.get(p) and f"her_{p}" in her_ids:
            for s in learned[p]:
                at = f"{DAY[s['day']]} {hour_label(s['hour'])}"
                rows.append({**s, "claim": claim(f"{at}: one of your best times on {LABEL[p]}, from your own results.", [f"her_{p}"], "her_data")})
        else:
            for s in slots[p]:
                at = f"{DAY[s['day']]} {hour_label(s['hour'])}"
                if p == "tiktok" and s["day"] in (0, 6):
                    c = claim(f"{at}: the big studies disagree on weekends, so this slot is a test your results will settle.", ["b_sprout_tt", "b_buffer_all"], "web", solid=False)
                elif p == "tiktok":
                    c = claim(f"{at}: weekday late afternoon and evening is where both big TikTok studies overlap.", ["b_sprout_tt", "b_buffer_all"], "web")
                elif p == "instagram":
                    c = claim(f"{at}: Instagram evenings and Tuesday to Wednesday are strongest across 9.6M posts; Thursday mornings also do well.", ["b_buffer_ig", "b_sprout_ig"], "web")
                else:
                    c = claim(f"{at}: YouTube Shorts did best Friday around 4 pm, with Saturday close behind.", ["b_buffer_all"], "web")
                rows.append({**s, "claim": c})
        out[p] = rows
    return out


# ---------- clean the model's answer ----------

def clean_claim(c: Any, ids: set[str]) -> dict[str, Any] | None:
    if not isinstance(c, dict) or not isinstance(c.get("text"), str) or not c["text"].strip():
        return None
    src = [s for s in (c.get("source_ids") or []) if isinstance(s, str) and s in ids]
    basis = c.get("basis") if c.get("basis") in ("her_data", "web", "upload") else "web"
    conf = "solid" if c.get("confidence") == "solid" and src else "uncertain"
    return {"text": c["text"].strip()[:600], "source_ids": src, "basis": basis, "confidence": conf}


def clean_list(xs: Any, ids: set[str], cap: int = 10) -> list[dict[str, Any]]:
    out = [clean_claim(c, ids) for c in (xs if isinstance(xs, list) else [])]
    return [c for c in out if c][:cap]


def clean_body(raw: dict[str, Any], ids: set[str], times: dict[str, Any]) -> dict[str, Any]:
    themes = []
    for t in raw.get("themes") or []:
        if isinstance(t, dict) and isinstance(t.get("title"), str):
            cl = clean_list(t.get("claims"), ids, 5)
            if cl:
                themes.append({"title": t["title"][:120], "claims": cl})
    creators = []
    for c in raw.get("comparable_creators") or []:
        if isinstance(c, dict) and isinstance(c.get("handle"), str) and c.get("platform") in PLATFORMS:
            why = clean_claim(c.get("why"), ids)
            if why:
                creators.append({"handle": c["handle"][:80], "platform": c["platform"], "why": why})
    return {
        "audience": clean_list(raw.get("audience"), ids),
        "themes": themes[:5],
        "hooks": clean_list(raw.get("hooks"), ids),
        "cut_styles": clean_list(raw.get("cut_styles"), ids),
        "best_times": times,
        "comparable_creators": creators[:6],
        "shot_list": clean_list(raw.get("shot_list"), ids, 8),
    }


# ---------- main ----------

def main(job: Job, spec: dict[str, Any]) -> dict[str, Any]:
    keys = spec.get("keys") or {}
    or_key = keys.get("openrouter") or os.environ.get("OPENROUTER_API_KEY")
    fc_key = keys.get("firecrawl") or os.environ.get("FIRECRAWL_API_KEY")
    if not or_key:
        raise RuntimeError("openrouter key missing")
    profile = spec.get("profile") or {}
    sources: list[dict[str, Any]] = [dict(s) for s in spec["baseline"]["sources"]]
    excerpts: dict[str, str] = {s["id"]: "Posting-time and frequency study used for the section 10b launch baseline." for s in sources}
    sources.append({"id": "her_profile", "url": None, "title": "Your locked Brand Profile", "kind": "her_data"})
    excerpts["her_profile"] = json.dumps(profile)[:6000]

    her = spec.get("her_data") or {}
    her_ids: set[str] = set()
    for p in PLATFORMS:
        s = (her.get("summary") or {}).get(p) or {}
        if s.get("videos"):
            sid = f"her_{p}"
            her_ids.add(sid)
            sources.append({"id": sid, "url": None, "title": f"Your {LABEL[p]} results ({s['videos']} videos)", "kind": "her_data"})
            excerpts[sid] = json.dumps({"summary": s, "accounts": [a for a in her.get("accounts") or [] if a.get("platform") == p], "timezone": her.get("timezone")})

    job.progress("reading uploads", 0, 3)
    work = WORK / "research"
    work.mkdir(parents=True, exist_ok=True)
    for u in spec.get("uploads") or []:
        local = work / u["id"]
        try:
            download_input(u["r2_key"], local)
            text, _ = extract_text(local, "", u.get("ext", ""))
            sources.append({"id": u["source_id"], "url": None, "title": u["title"][:200], "kind": "upload"})
            excerpts[u["source_id"]] = text[: EXCERPT * 3]
        except (Unreadable, Exception):  # noqa: BLE001
            log("research.upload.unreadable")
        finally:
            local.unlink(missing_ok=True)
    log("research.uploads", count=sum(1 for s in sources if s["kind"] == "upload"))

    job.progress("searching the web", 1, 3)
    if fc_key:
        seen: set[str] = set()
        n = 0
        for q in queries_from_profile(profile):
            for r in firecrawl_search(fc_key, q):
                if r["url"] in seen:
                    continue
                seen.add(r["url"])
                n += 1
                sid = f"w{n}"
                sources.append({"id": sid, "url": r["url"], "title": r["title"], "kind": "web"})
                excerpts[sid] = r["text"]
        log("research.web", sources=n)
    else:
        sources.append({"id": spec.get("web_skipped_source_id", "web_skipped"), "url": None, "title": "Web search was skipped: Firecrawl is not connected", "kind": "web"})
        log("research.web.skipped")

    if (spec.get("features") or {}).get("deeper_research"):
        text, pages = deeper_search(or_key, spec.get("deeper_model", "perplexity/sonar"), profile)
        for i, pg in enumerate(pages):
            sid = f"d{i + 1}"
            sources.append({"id": sid, "url": pg["url"], "title": pg["title"], "kind": "web"})
            excerpts[sid] = text if i == 0 else "Cited by the deeper search summary above."
        log("research.deeper", sources=len(pages))

    job.progress("writing the brief", 2, 3)
    usable = [s for s in sources if s["id"] != spec.get("web_skipped_source_id", "web_skipped")]
    ids = {s["id"] for s in usable}
    listing = "\n\n".join(f"[{s['id']}] ({s['kind']}) {s['title']}\n{excerpts.get(s['id'], '')[:EXCERPT]}" for s in usable)
    user = f"BRAND PROFILE:\n{json.dumps(profile)[:6000]}\n\nSOURCES (cite by id):\n{listing}"
    raw = llm_json(or_key, spec["model"], spec["system"], user[:120_000], max_tokens=4000)
    body = clean_body(raw, ids, best_times(spec, her_ids))
    if not body["audience"] and not body["themes"]:
        raise RuntimeError("draft came back empty")
    job.progress("writing the brief", 3, 3)
    log("research.done", sources=len(sources), themes=len(body["themes"]))
    return {"body": body, "sources": sources}


if __name__ == "__main__":
    run(main)

File: worker/jobs/research.ts
// Research Brief from her data + web + her uploads (BUILD_PLAN.md section 6). jobs/research.py
// fetches this spec, searches the web (jobs/common.py Web: Firecrawl when she connected it, else
// the free keyless search, so it never needs a key), reads her uploaded reports, and asks an OpenRouter free model to draft a
// BriefBody (shared/types.ts) citing only the sources it was given. The Worker then enforces
// the truth rules (domain/brief.ts) before storing a new draft version.
import type { JobHandler } from "./registry";
import type { Env } from "../env";
import { DEFAULT_FEATURES, DEFAULT_WEEKLY_CAPS, LAUNCH_SLOTS, PLATFORMS, type Platform } from "@shared/constants";
import type { BriefBody, BriefSource, Features } from "@shared/types";
import { getConnectionSecret } from "../lib/connections";
import { getSetting, parseJson, recordEvent } from "../lib/db";
import { log } from "../lib/log";
import { BASELINE_SOURCES, enforceTruth, shapeProblems, sourceProblems } from "../domain/brief";
import { bucketByTime, historyDays } from "../domain/learning";
import { FREE_MODEL } from "../services/openrouter";
import { readSettings } from "../routes/settings";
import { loadObservations } from "./metrics";
import { buildFakeBrief } from "./research_fake";

export const RESEARCH_SYSTEM = `You write a Research Brief for a short-form video creator. Answer with ONE JSON object only.
Keys: "audience": Claim[], "themes": [{"title": string, "claims": Claim[]}] (3 to 5 themes), "hooks": Claim[] (hook formulas in her voice),
"cut_styles": Claim[] (cut styles and lengths to favor), "comparable_creators": [{"handle": string, "platform": "tiktok"|"instagram"|"youtube", "why": Claim}],
"shot_list": Claim[] (5 to 8 things to film next).
Claim = {"text": string, "source_ids": string[], "basis": "her_data"|"web"|"upload", "confidence": "solid"|"uncertain"}.
TRUTH RULES: cite only source ids from the SOURCES list. Every claim cites at least one source unless it is marked "uncertain".
basis is "her_data" for her profile or her own numbers, "web" for web sources, "upload" for reports she uploaded.
Weak, indirect or conflicting evidence is "uncertain". Never invent numbers, creators or studies. Name a creator only if a web source names them.
Plain words, short sentences.`;

export interface ResearchResult {
  body: BriefBody;
  sources: BriefSource[];
}

async function herData(env: Env) {
  const s = await readSettings(env);
  const obs = await loadObservations(env);
  const buckets = bucketByTime(obs, s.audience_timezone);
  const summary = {} as Record<Platform, { videos: number; avg_views: number; history_days: number; top_times: { day: number; hour: number; avg_views: number; posts: number }[] }>;
  for (const p of PLATFORMS) {
    const mine = obs.filter((o) => o.platform === p);
    summary[p] = {
      videos: mine.length,
      avg_views: mine.length ? Math.round(mine.reduce((a, o) => a + o.views, 0) / mine.length) : 0,
      history_days: historyDays(obs, p),
      top_times: buckets.filter((b) => b.platform === p).slice(0, 5).map(({ day, hour, avg_views, posts }) => ({ day, hour, avg_views, posts })),
    };
  }
  const { results: accounts } = await env.DB.prepare(
    "SELECT a.platform, a.followers, a.avg_views, a.captured_at, a.source FROM account_stats a WHERE a.captured_at = (SELECT MAX(b.captured_at) FROM account_stats b WHERE b.platform = a.platform)",
  ).all<{ platform: Platform; followers: number; avg_views: number; captured_at: string; source: string }>();
  return { summary, accounts, learned_slots: await getSetting(env.DB, "learned_slots", {}), timezone: s.audience_timezone };
}

async function buildSpec(env: Env, jobId: string) {
  const profile = await env.DB.prepare("SELECT sections FROM brand_profile WHERE locked = 1 ORDER BY version DESC LIMIT 1").first<{ sections: string }>();
  const { results: uploads } = await env.DB.prepare("SELECT id, file_name, r2_key FROM research_uploads ORDER BY uploaded_at DESC LIMIT 10").all<{ id: string; file_name: string; r2_key: string }>();
  const present: typeof uploads = [];
  for (const u of uploads) if (await env.FILES.head(u.r2_key)) present.push(u);
  const features = await getSetting<Features>(env.DB, "features", { ...DEFAULT_FEATURES });
  return {
    job_id: jobId,
    type: "research",
    profile: parseJson(profile?.sections, null),
    her_data: await herData(env),
    uploads: present.map((u) => ({ id: u.id, source_id: `up_${u.id}`, title: u.file_name, r2_key: u.r2_key, ext: (u.file_name.split(".").pop() ?? "").toLowerCase().slice(0, 5) })),
    baseline: { basis: "BUILD_PLAN.md section 10b", slots: LAUNCH_SLOTS, caps: DEFAULT_WEEKLY_CAPS, sources: BASELINE_SOURCES },
    features: { deeper_research: !!features.deeper_research },
    // Firecrawl is optional: without it the job searches with the free keyless path (jobs/common.py Web).
    keys: { openrouter: await getConnectionSecret(env, "openrouter"), firecrawl: await getConnectionSecret(env, "firecrawl") },
    model: FREE_MODEL,
    deeper_model: "perplexity/sonar",
    system: RESEARCH_SYSTEM,
  };
}

export const researchJob: JobHandler = {
  async buildSpec(env, jobId) {
    return buildSpec(env, jobId);
  },

  async applyResult(env, jobId, refId, raw) {
    void refId;
    const r = (raw ?? {}) as Partial<ResearchResult>;
    const problems = [...sourceProblems(r.sources), ...shapeProblems(r.body)];
    if (problems.length) {
      log.error("research.apply.invalid", { problems: problems.length });
      throw new Error(`brief failed shape check (${problems.length} problems)`);
    }
    const sources = r.sources as BriefSource[];
    const body = enforceTruth(r.body as BriefBody, sources);
    await env.DB.prepare("UPDATE research_briefs SET status = 'superseded' WHERE status = 'draft'").run();
    const res = await env.DB.prepare("INSERT INTO research_briefs (body, sources, status) VALUES (?, ?, 'draft')").bind(JSON.stringify(body), JSON.stringify(sources)).run();
    await recordEvent(env.DB, "brief.drafted", jobId, { version: res.meta.last_row_id, sources: sources.length });
    log.info("research.apply", { sources: sources.length });
  },

  async onFailure(env, jobId, refId, safeError) {
    void refId;
    await recordEvent(env.DB, "brief.failed", jobId, { len: safeError.length });
    log.warn("research.failed");
  },

  async fakeRun(env, jobId) {
    const spec = await buildSpec(env, jobId);
    const stats: Partial<Record<Platform, { videos: number }>> = {};
    for (const p of PLATFORMS) stats[p] = { videos: spec.her_data.summary[p].videos };
    return buildFakeBrief({ stats, uploads: spec.uploads.map((u) => ({ id: u.id, title: u.title })) });
  },
};
